### Задание 1
Импортируйте библиотеки pandas и numpy.

In [10]:
import numpy as np
import pandas as pd

Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn. Создайте датафреймы X и y из этих данных.

In [13]:
from sklearn.datasets import load_boston
boston = load_boston()
data = boston.data
feature_names = boston.feature_names
target = boston.target
x = pd.DataFrame(data, columns=feature_names)
y = pd.DataFrame(target, columns=['price'])

Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.

In [14]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.3, random_state=42)

Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.

In [15]:
from sklearn.linear_model import LinearRegression

In [16]:
lr = LinearRegression()


Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.

In [17]:
lr.fit(x_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [18]:
y_pred = lr.predict(x_test)

In [19]:
check_test = pd.DataFrame({'y_test': y_test['price'],
                          'y_pred':y_pred.flatten()},
                         columns=['y_test', 'y_pred'])

In [20]:
check_test.head(10)

y_test     y_pred
173    23.6  28.648960
274    32.4  36.495014
491    13.6  15.411193
72     22.8  25.403213
452    16.1  18.855280
76     20.0  23.146689
316    17.8  17.392124
140    14.0  14.078599
471    19.6  23.036927
500    16.8  20.599433

Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.

In [21]:
from sklearn.metrics import r2_score
r2_score(y_test, y_pred)

0.711226005748496

### Задание 2
Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble.
Сделайте агрумент n_estimators равным 1000, max_depth должен быть равен 12 и random_state сделайте равным 42.


In [23]:
from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor(n_estimators=100, max_depth=2, random_state=0)

Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression, но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0], чтобы получить из датафрейма одномерный массив Numpy,так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.



In [24]:
model.fit(x_train, y_train.values[:, 0])

RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=2,
                      max_features='auto', max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=1, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, n_estimators=100,
                      n_jobs=None, oob_score=False, random_state=0, verbose=0,
                      warm_start=False)

Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего задания.

In [25]:
y_pred_RFR = model.predict(x_test)

In [26]:
check_test_RFR = pd.DataFrame({'y_test': y_test['price'],
                          'y_pred_RFR':y_pred_RFR.flatten()},
                         columns=['y_test', 'y_pred_RFR'])

In [27]:
check_test_RFR.head(10)

y_test  y_pred_RFR
173    23.6   22.688943
274    32.4   28.558295
491    13.6   15.448460
72     22.8   25.826828
452    16.1   15.539578
76     20.0   22.069080
316    17.8   15.601542
140    14.0   13.937275
471    19.6   22.007116
500    16.8   22.069080

In [28]:
r2_score(y_test, y_pred_RFR)

0.744105678697021

R2 показатель у RandomForestRegressor (0.744105678697021) выше чем у LinearRegression (0.711226005748496). Соответственно RandomForestRegressor работает для данного датасета лучше